### PR TITLE
One hash-based mechanism for "shipped default vs user-customized"

### LINKED ISSUE #401: One hash-based mechanism for 'shipped default vs user-customized' across prompts, skills and tool descriptions
## The problem

The plugin ships editable text in several places — base system prompt, memory prompt, bundled skill bodies, tool descriptions — and every one of them needs the same question answered: **is what is on disk still something we shipped, or did the user change it?**

The answer decides whether an update can be applied silently or must be surfaced as a notice. There are currently **three independent implementations, one of which is missing entirely**, and they behave differently.

### 1. Base system prompt — full-text map, works

`src/agent/prompts.ts`:

```ts
/**
 * Maps each historical version number to the exact prompt string shipped at that version.
 * Lets normalizeAgent() detect "still on old default" vs "user customized" without storing
 * a hash. Keep all entries forever — never remove old versions.
 */
export const HISTORICAL_SYSTEM_PROMPTS: ReadonlyMap<number, string> = new Map([[1, BASE_SYSTEM_PROMPT]]);
```

Consumed twice: auto-migration (`dataStore.svelte.ts:482`) and staleness detection (`detectStaleGuidance`, `:2330`) which feeds the dismissable notice + diff in `chatRecommendations.ts`.

### 2. Memory prompt — no history at all

`dataStore.svelte.ts:518`:

```ts
const isShippedDefault = !oldPrompt.trim() || oldPrompt === DEFAULT_MEMORY_PROMPT;
```

Compared against the **current** constant only. There is no `HISTORICAL_MEMORY_PROMPTS`. So the moment `DEFAULT_MEMORY_PROMPT` changes, every existing install's memory prompt is classified as user-customized — it can never be auto-migrated, and it never raises a staleness notice either. It just silently diverges forever.

### 3. Bundled skills — existence-only

`SkillsService.bootstrapDefaultSkills`:

```ts
// Skip if already exists (user may have customized)
if (await this.adapter.exists(`${skillsDir}/${bundledSkill.name}/${SKILL_FILENAME}`)) continue;
```

Nothing reads `metadata.version` — the field is decorative. A fast path above returns early when all seed folders exist, so a bumped version never reaches the per-skill check. **An improvement to a bundled core skill reaches new vaults only.**

Hit concretely in #399: `explore-vault` gained reformulation guidance and went `1.0` → `1.1`; the test vault stayed on `1.0`. Only remedy is deleting the folder by hand.

### 4. Tool descriptions — ad-hoc set (added in #399)

`SEARCH_NOTES_DESC_DEFAULTS` is a `Set<string>` of the two shipped variants, so `createSearchNotesTool` can swap between them without clobbering a customization. Same idea again, fourth spelling.

## Proposal: one hash-based mechanism for all of them

Every consumer above only ever does an **equality/membership test** — `.includes(oldPrompt)`, `historical === content`, `SET.has(description)`. None of them reads the historical text back. So storing full text buys nothing over a hash.

```ts
/** version → sha256 of the exact text shipped at that version. */
type ShippedHistory = ReadonlyMap<number, string>;

/** Did we ship this exact content at some point? */
function isShippedDefault(content: string, history: ShippedHistory): boolean;

/** Which version, if any — lets a notice say "your copy is from v2". */
function shippedVersion(content: string, history: ShippedHistory): number | null;
```

Applied uniformly to base prompt, memory prompt, bundled skill bodies and tool descriptions.

**Why hashing matters more for skills than prompts:** skills are bundled *files* and there are ~10 of them, so retaining every historical body as a string constant grows the bundle with text that is never displayed. For prompts the saving is smaller but the consistency is the point — one mechanism, one set of semantics, one place to get the normalization right.

### The trade-off, stated honestly

A hash cannot reconstruct old text, so a three-way *"old default vs new default vs yours"* diff becomes impossible. This looks acceptable: the diff a user actually wants is **"your version vs the current default"**, and both sides of that are available at runtime. If a three-way diff is wanted later, the old text would have to be retained deliberately — worth deciding now rather than discovering the constraint afterwards.

## Behaviour once unified

For each surface, on load:

- hash **matches current shipped version** → nothing to do
- hash **matches an older shipped version** → untouched old default, update silently
- **no match** → user edited it. Leave it alone, and emit a `StaleGuidance` record so the existing notice + diff surfaces "the default moved"

`StaleGuidance.kind` is currently the single literal `"system-prompt"`. Adding `"memory-prompt"` and `"skill"` reuses the existing notice, diff and dismissal path — `chatRecommendations.ts` already renders whatever is in the list.

## Notes

- **Normalize before hashing** — line endings and a trailing newline must not read as a customization. Getting this right once, centrally, is a large part of the value.
- `HISTORICAL_SYSTEM_PROMPTS` currently holds exactly one entry, `[1, BASE_SYSTEM_PROMPT]`, aliasing the live constant. No genuinely historical text is stored yet, so **converting it costs nothing today** — the migration is cheapest now and gets more expensive with every shipped version.
- Bundled `SKILL.md` `metadata.version` is unread. Either wire it up or drop it; a version field that does nothing invites exactly the confusion that produced this issue.
- Only affects **shipped** content. User-created skills have no shipped version and are never touched.
- Precedent for a cruder first step: `migrateCoreSkills` (schema v6) deleted orphaned `GUIDANCE.md` files so core skills re-seeded cleanly. A schema bump could force re-seed of stale-but-unmodified core skills today with no new machinery — worth doing if the full unification is deferred.

## Impact until fixed

`explore-vault`'s reformulation guidance from #399 ships to new vaults only. That guidance is what the `reformulation` benchmark tier was built to justify, so it is currently unreachable for existing installs. The memory-prompt gap is quieter but worse in kind: it will misclassify silently, with no notice, the first time that default changes.

### PR DESCRIPTION
**Pull request by Leo310**
Resolves #401.

## The problem

Four places ship editable text and each answered "is what's on disk still something we shipped, or did the user change it?" differently. Two were broken:

- **Memory prompt** — no history at all. Compared against the *current* constant only, so the first change to `DEFAULT_MEMORY_PROMPT` would classify every existing install as user-customized: never auto-migrated, and never flagged either. Silent divergence.
- **Bundled skills** — existence-only. `metadata.version` was decorative, and a fast path returned early once all seed folders existed, so a bumped version never reached the per-skill check. Improvements reached new vaults only — hit concretely when `explore-vault` went `1.0 → 1.1` in #399.

## The mechanism

New `src/utils/shippedDefaults.ts`: a `ShippedHistory` (version → fingerprint) plus `isShippedDefault` / `shippedVersion`, with normalization applied centrally. Per surface, on load:

- matches the **current** shipped version → nothing to do
- matches an **older** shipped version → untouched old default, update silently
- **no match** → the user edited it. Leave it alone, emit a `StaleGuidance` record

**Normalization is the load-bearing part.** Line endings and a trailing newline must not read as a customization — content round-trips through the vault adapter and whatever editor the user opens the note in, and either can change them without the user touching a character. Getting that wrong misclassifies installs wholesale, in the direction that silently withholds updates.

### Three design decisions worth reviewing

**FNV-1a, not sha256.** `node:crypto` is externalized in `vite.config.ts` via `builtinModules` and unavailable on mobile (`isDesktopOnly: false`), and `crypto.subtle.digest` is async — while `data.staleGuidance` is a *synchronous* Svelte reactive getter read during render. Async would mean a precomputed cache plus `$state` invalidation for no functional gain. Collision resistance is irrelevant here: this is change detection over a handful of strings we shipped ourselves.

**Fingerprints, not full text.** The 10 bundled `SKILL.md` files are ~67 KB today (`bases` 21 KB, `canvas` 14 KB). Retaining every revision verbatim would grow the bundle with text that is never displayed. Every consumer only ever does an equality test — none reads historical text back.

*Trade-off, stated deliberately:* a fingerprint can't reconstruct old text, so a three-way "old default vs new default vs yours" diff becomes impossible. This costs nothing today — `openSystemPromptDiff` already passes only `getPrompt` + `defaultPrompt`, so the diff is already two-way.

**Tool descriptions are excluded, not unified.** They aren't user-editable — `ToolConfigForm` says so outright (*"name/description aren't editable in this form"*) and renders no input. A stored description is therefore only ever a cached shipped default. The `Set` membership tests were never customization detection; they existed so the *automatic* variant swap (`read_content` by processor capability, `search_notes` by embedding index) didn't freeze at whichever variant was current on load. Both sets and their guards are deleted in favour of unconditional recompute — net deletion.

## Skills

- Three-way reconcile in `bootstrapDefaultSkills`; the early-return fast path is removed (it's precisely what made a bumped version unreachable).
- `seedIntegrationSkill` had the **same skip-if-exists gap one level down**. Community integration skills seed on demand, so they're absent from the startup seed set — reconciling only that set would leave them written once and never re-checked. Bootstrap now also reconciles bundled skills already installed but outside the seed set, while deliberately *not* installing ones the user hasn't opted into.
- `metadata.version` is now load-bearing (#401's "wire it up or drop it" → wire it up), with a test asserting every bundled skill declares one, since omitting it silently opts a skill out of updates.

**No backfill.** With no active users there's nothing to preserve, so `explore-vault` is reset to `1.0` and every bundled skill's current body is its sole shipped version. `PRIOR_SKILL_BODIES` is an empty, documented seam — the previous body gets pasted there when a skill is next revised. Skipping that step is the one way to reintroduce this bug.

## Also

- `StaleGuidance.kind` gains `"memory-prompt"` and `"skill"`, reusing the existing notice / diff / dismissal path — `filterUpdateNotices` needed no change.
- `updateNoticeId` includes the skill name; otherwise dismissing one skill's notice would dismiss every skill's.
- `AgentEditorModal`'s drift checks used `.trim()`, weaker than the shared normalization — a CRLF round-trip would have offered a "Diff with default" button against identical text.

## Verification

`bun run test` (1347 passing), `bun run check` clean, `bun run build` succeeds.

New coverage: the normalization contract (CRLF / trailing-newline match, interior whitespace does *not*); the three-way skill outcome driven by a synthetic prior version — which also exercises the seam future revisions use; both integration-skill directions; and an unreadable file being flagged rather than clobbered.

**One behavioural note:** bootstrap now costs one `read()` per skill folder instead of `exists()`. Bounded (~10 folders) and inside `StartupProfiler.measure("skills:bootstrap")`, so any regression surfaces in the startup profile.

_AI assistance: written with AI coding agents from the maintainer's brief; reviewed and tested by the maintainer._

### CODE REVIEW

File: src/utils/shippedDefaults.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+ * file that differs from what we shipped only by line endings or a trailing newline is NOT
+ * a customization, and treating it as one would misclassify installs wholesale — silently,
+ * and in the direction that withholds updates.
+ *
+ * Both cases are routine rather than hypothetical: content round-trips through the vault
+ * adapter and through whatever editor the user opens the note in, and either can normalize
+ * line endings or append a final newline without the user touching a character.
+ */
+export function normalizeShipped(text: string): string {
+	return text.replace(/\r\n/g, "\n").trim();
```
[the remark this answers is not in the document]
Comment (Leo310) [reply src/utils/shippedDefaults.ts:47]: Declining this one deliberately — the tradeoff runs the other way.

For this to cause harm, a user's **sole** edit to a prompt or skill would have to be adding blank lines or trailing spaces at the file boundary, leaving every interior byte identical. That isn't a plausible customization, and it carries no meaning to the model: these fragments get concatenated during prompt assembly, so boundary whitespace is discarded downstream regardless.

The failure it guards against is common and silent, in the opposite direction. These files round-trip through the vault adapter and through whatever editor the user opens the note in, and either can normalize line endings or append a final newline **without the user touching a character**. Without boundary normalization an untouched file reads as customized — so it silently stops receiving updates *and* nags the user about a change they never made. That's the exact class of bug #401 exists to fix, and it would misclassify installs wholesale rather than in a contrived edge case.

Interior whitespace is deliberately preserved — reflowing a paragraph or re-indenting a list is a real edit. `does NOT ignore interior whitespace changes` in `test/utils/shippedDefaults.test.ts` pins that boundary, so the trimming is scoped rather than broad.

Net: the current semantics risk discarding an edit nobody makes; the proposed change breaks updates for edits nobody made.
